### PR TITLE
Generate JavaKitFunction module

### PR DESCRIPTION
**Pull request by iainsmith**
Add JavaKitFunction module & Makefile action as first steps for #84 

## Java
```java
public Predicate<Integer> lessThanTen() {
  Predicate<Integer> predicate = i -> (i < 10);
  return predicate;
}
```

## Swift

```swift
let predicate: JavaPredicate<JavaInteger> = self.lessThanTen()! 
let value = predicate.test(JavaInteger(3).as(JavaObject.self)) // question about removing need for this cast below?
print("Running a predicate from swift 3 < 10 = \(value)")
```

Updated JavaKitExample available in this commit https://github.com/swiftlang/swift-java/commit/7baa992298f423b06145562f4bbf72a381a63ca1

### Questions

- Is this roughly what you had in mind or do you want to go with a different direction?
- For the `callAsFunction` implementation, should we add those by hand or is it worth updating the generator / config?
- Are you happy if we update the existing JavaKitExample to import `JavaKitCollection` and `JavaKitFunction` so these get run through CI? 
- The generated API doesn't enforce the Generic parameter. Should we update the `@JavaMethod` macro to support handling the cast to JavaObject? That would let us move from: 

```diff
@JavaInterface("java.util.function.Predicate")
public struct JavaPredicate<T: AnyJavaObject> {
  @JavaMethod
-  public func test(_ arg0: JavaObject?) -> Bool
+  public func test(_ arg0: T?) -> Bool

// Callsite

let predicate: JavaPredicate<JavaInteger> = self.lessThanTen()! 
- let value = predicate.test(JavaInteger(3).as(JavaObject.self))
+ let value = predicate.test(JavaInteger(3))
```

### CODE REVIEW

**Comment by ktoso** [comment]
Looks fine, PR doesn't seem to include changes to samples or a test though.

Would you be interested in helping out in the other direction as well? We need to support all these in jextract as well: https://github.com/swiftlang/swift-java/issues/101

**Comment by iainsmith** [comment]
> PR doesn't seem to include changes to samples or a test though.

- Do you want a new sample or to add a few functions to the existing sample in JavaKitExample/com/example/swift/HelloSwift.java?
- What's the approach for testing generated modules like these, I couldn't see anything for JavaKitCollection?

> We need to support all these in jextract as well: https://github.com/swiftlang/swift-java/issues/101

I haven't been able to get `make jextract-run` to run yet, using either `main-snapshot-2024-10-25` or `6.0-snapshot-2024-10-12`, but I'd be keen to give it a go as time allows.

File: Samples/JavaKitSampleApp/Sources/JavaKitExample/JavaKitExample.swift [hunk 1]
```diff
@@ -44,6 +45,10 @@ extension HelloSwift: HelloSwiftNativeMethods {
     self.name = "a 🗑️-collected language"
     _ = self.sayHelloBack(42)
 
+    let predicate: JavaPredicate<JavaInteger> = self.lessThanTen()!
+    let value = predicate.test(JavaInteger(3).as(JavaObject.self))
+    print("Running a JavaPredicate from swift 3 < 10 = \(value)")
```
Comment (iainsmith) [inline Samples/JavaKitSampleApp/Sources/JavaKitExample/JavaKitExample.swift:50]: Not a particular compelling example, but illustrates the module working.

Comment (ktoso) [reply Samples/JavaKitSampleApp/Sources/JavaKitExample/JavaKitExample.swift:50]: I think this is fine, just at least some soundness check that it works at all 👍 

I'd put it into a test file in this project and we're good them IMHO.

**Review by ktoso — APPROVED** [review]

**Pull request merged**